geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/jobs/PartitionOutputFormat.scala

```diff
@@ -51,49 +54,135 @@ class PartitionOutputFormat(delegate: SingleFileOutputFormat) extends OutputForm
+        val geom = value.getDefaultGeometry.asInstanceOf[Geometry]
+        if (geom != null) {
+          bounds.expandToInclude(geom.getEnvelopeInternal)
+        }
+        count += 1L
+      }
+
+      def meta: PartitionMetadata = {
+        val f = files.map(StorageFile(_, System.currentTimeMillis()))
+        PartitionMetadata(partition, f, PartitionBounds(bounds), count)
+      }
+
+      def close(context: TaskAttemptContext): Unit =
+        context.getCounter(Group, StorageConfiguration.Counters.partition(partition)).increment(count)
+
+      protected def newWriter(): (Path, RecordWriter[Void, SimpleFeature]) = {
+        val file = StorageUtils.nextFile(workPath, partition, leaf, encoding, fileType)
+        logger.debug(s"Creating record writer at path $file")
+        files += file.getName
+        // noinspection LanguageFeature
+        (file, delegate.getRecordWriter(context, file))
       }
-      CloseWithLogging(metadata)
     }
 
-    private def createWriter(partition: String): PartitionState = {
-      val root = delegate.getOutputCommitter(context).asInstanceOf[FileOutputCommitter].getWorkPath
-      // TODO combine this with the same code in ParquetFileSystemStorage
-      val file = StorageUtils.nextFile(root, partition, metadata.leafStorage, metadata.encoding, fileType)
-      logger.debug(s"Creating record writer at path $file")
-      // noinspection LanguageFeature
-      PartitionState(root, file.getName, delegate.getRecordWriter(context, file))
+    private class SinglePartitionState(partition: String) extends PartitionState(partition) {
+
+      private val writer = newWriter()._2
+
+      override def write(key: Void, value: SimpleFeature): Unit = {
+        writer.write(key, value)
+        super.write(key, value)
+      }
+
+      override def close(context: TaskAttemptContext): Unit = {
+        writer.close(context)
+        super.close(context)
+      }
+    }
+
+    private class ChunkedPartitionState(partition: String, estimator: FileSizeEstimator, context: TaskAttemptContext)
+        extends PartitionState(partition) {
+
+      private var count = 0L // number of features written
+      private var total = 0L // sum size of all finished chunks
+      private var remaining = estimator.estimate(0L)
+
+      private var path: Path = _
+      private var writer: RecordWriter[Void, SimpleFeature] = _
+
+      override def write(key: Void, value: SimpleFeature): Unit = {
+        if (writer == null) {
+          val (p, w) = newWriter()
+          path = p
+          writer = w
+        }
+        writer.write(key, value)
+        count += 1
+        remaining -= 1
+        if (remaining == 0) {
+          writer.close(context)
+          writer = null
+          // adjust our estimate to account for the actual bytes written
+          total += storage.context.fc.getFileStatus(path).getLen
+          estimator.update(total, count)
+          remaining = estimator.estimate(0L)
+        }
```

Comment:
I'm only skimming.  Seems like this logic is in the ChunkedFileWriter as well?  Is there a way to refactor it?  (Says the guy who cannot quite find things anymore after all the clever refactoring.)

Reply:
it is very similar, but there are enough differences in how the paths are generated, the writers are closed (the output format requires `close(context)`), etc that i didn't think it was worthwhile to wrap a new trait around the two. the logic is also not terribly complex, although not entirely trivial either. i could be persuaded to unify it if you feel strongly...

Reply:
Since the code is slightly different, I don't feel terribly strongly about a unification.  (If there were a trivial refactoring, yeah, maybe....)

Seems reasonable to leave it as is!